### DISCUSSION TITLE
Tracer lifetime

### DISCUSSION THREAD

**Discussion by demurgos** [open-telemetry/opentelemetry-specification#4060] [Q&A]
Hello, I am adding OpenTelemetry instrumentation to a Rust HTTP server. I am seeking clarification about the recommended lifetime of a `Tracer` instance. In particular I am trying to decide between:
1. the middleware/tracing layer should instantiate a single long-lived `Tracer` instance (call `TracerProvider.tracer` when creating the layer)
2. a new `Tracer` should be acquired for each request

My understanding is that the first approach is better as this middleware layer represents a single instrumentation scope. I also understand that both ways should work, but I am trying to keep my instrumentation idiomatic.

The documentation for `TracerProvider` currently has the following sentence which clarifies that it usually has the lifetime of the application:
> In most applications, a Tracer Provider is initialized once and its lifecycle matches the application’s lifecycle.

Unfortunately the `Tracer` documentation does not have such guidance.

What tracer lifetime would you recommend? Instantiate once for the middleware layer, or on each request?

**Comment by dmathieu** [comment]
In terms of functionality, both work.
However, in terms of performance, usually, the second option is better, as retrieving a global tracer stored in memory is heavier than keeping one around within an object.

This may be a rust specific question though.

**Reply by demurgos** [reply]
I assume that you mean the first option (create the tracer with the middleware layer).

I am currently adding OpenTelemetry to a Rust project, but we also deploy Node, Java and PHP services so I was looking for general guidance as I'll need to convert those too.

Getting the tracer only once when creating the middleware is what I'm using currently since it feels more sensible, so I'll keep it this way. Thank you for your feedback.
